lib/settings_validator.rb

```diff
@@ -0,0 +1,317 @@
+    %i(yml exception_notifier exception_recipients) =>
+      [:validate_value_has_format_email],
+    %i(yml allowed_iri_schemes) => [:validate_value_is_present],
+    %i(yml formality_levels) => [:validate_value_has_keys, :name, :description],
+    %i(yml license_models) => [:validate_value_has_keys, :name, :url],
+    %i(yml ontology_types) =>
+      [:validate_value_has_keys, :name, :description, :documentation],
+    %i(yml tasks) => [:validate_value_has_keys, :name, :description],
+
+    # hets.yml
+    %i(yml hets_path) => [:validate_value_is_absolute_filepath],
+    %i(yml hets_lib) => [:validate_value_is_absolute_filepath],
+    %i(yml hets_owl_tools) => [:validate_value_is_absolute_filepath],
+    %i(yml cmd_line_options) => [:validate_value_is_present],
+    %i(yml server_options) => [:validate_value_is_present],
+  }
+
+  attr_reader :errors
+
+  def initialize(config)
+    @config = config
+  end
+
+  def validate
+    @errors = []
+
+    HAS_TO_BE_PRESENT.each { |key_chain| validate_presence(key_chain) }
+    HAS_TO_BE_URL.each { |key_chain| validate_format_url(key_chain) }
+    HAS_TO_BE_EMAIL.each { |key_chain| validate_format_email(key_chain) }
+    HAS_TO_HAVE_TYPE.each do |types, key_chains|
+      key_chains.each { |key_chain| validate_type(key_chain, types) }
+    end
+
+    if Rails.env.production?
+      HAS_TO_BE_DIRECTORY_IN_PRODUCTION.each do |key_chain|
+        validate_directory_exists(key_chain)
+      end
+      HAS_TO_BE_PRESENT_IN_PRODUCTION.each do |key_chain|
+        validate_presence(key_chain)
+      end
+    end
+
+    validate_arrays
+    validate_specials
+
+    if @errors.present?
+      raise ValidationError.new(@errors.map(&:message).join("\n"))
+    end
+  end
+
+  protected
+
+  def validate_arrays
+    ARRAY_VALIDATIONS.each do |key_chain, call|
+      if array = get_value(key_chain)
+        array.each do |value|
+          method = call.first
+          args = call[1..-1]
+          send(method, key_chain, value, *args)
+        end
+      end
+    end
+  end
+
+  def validate_specials
+    validate_hets_path
+  end
+
+  def validate_hets_path
+    key_chain = %i(yml hets_path)
+    hets_paths = get_value(key_chain)
+    unless hets_paths.any? { |path| File.executable?(path) }
+      @errors << ResourceNotFound.new(key_chain,
+        "Executable 'hets' not found in any of the given paths: #{hets_paths}")
+    end
+  end
+
+  def get_value(key_chain)
+    object = base(key_chain.first)
+    key_chain[1..-1].each do |key|
+      object = object.send(key)
+      if object.nil?
+        @errors << KeyNotSet.new(key_chain, 'Key not set.')
+        return
+      end
+    end
+    object
+  end
+
+  def base(namespace)
+    if namespace == :yml
+      Settings
+    else
+      @config
+    end
+  end
+
+  def validate_with(key_chain, &block)
+    if value = get_value(key_chain)
+      block.call(value)
+    end
+  end
+
+  def validate_directory_exists(key_chain)
```

Comment:
What do you mean by meta data validation?

> where is the place to put the defaults?

The places for reasonable defaults are the non-local settings files, i.e. `settings.yml`, `environment/#{envronment}.yml`. Those need to be overwritten (adjusted to the specific instance) in the `settings.local.yml` and similar.

> how can ranges for certain values be described (e.g. min max

Currently, that's not possible. Could you give me a pointer to settings that need it?

> how can be system properties access? E.g. to define a reasonable default like #{numcpus} * 0.75

Same question.

> for those few settings a simple check at a central place like already shown in the1170-settings-cleanup.patch would have saved you probably a lot of time

As stated in the pull request, that's not the place for validations or setting defaults. (according to rails). That's the place for reading the settings and configuring some rails-internals. 

> So, forking a different gem, writing about 300 lines of code, which doesn't provide, what is actually needed, introducing other errors/potential inconsistencies is for my taste a little bit too much.

There are no solutions for the settings-validation task out there, so we needed to create one ourselves. I know this is not the most elegant one, but it is a central place for validation and for validation only. I would have liked to use `ActiveModel::Validations` for the validations because it already defines some neat validation-methods. After playing around a little bit with this class, I figured out that it needs a lot more code than what this pull request contains.

The fork is not a that big a deal. The important changes are in a pull request of the original gem anyway. It just needs to be reviewed and merged. Then we can use the original rails_config again.